### PR TITLE
Initial implementation

### PR DESCRIPTION
**Pull request by phamann**
### TL;DR
This is the initial port and working implementation of a Fastly provider for [Open Insights](https://github.com/openinsights/openinsights) which is a port of the current [Insights.js](https://github.com/fastly/insights.js) logic. I.e. Open Insights contains the core runtime which initialises and orchestrates the task running sessions on a  page load and the "provider" encapsulates all of the Fastly specific logic to fetch task configuration from our API and how to translate fetch data from the framework to our own custom schema. 

This has been successfully running as a canary for many months in production and we are now confident to promote it to 100% of traffic and open-source the components.

### Notes for reviewers
- No core functionality has been added, or business logic changed, from Insights.js - we've just restructured it to implement the Open Insights Provider Class.
- Everything inside the `src/util` and `src/lib` directories are a 1:1 copy and thus don't really need to be reviewed, though may need to be  read to understand the libraries control flow.
- The `src/index.js` and `src/index.spec.js` are the two new files and are what we're keen for most of the review to center around.
- To run the code, you'll need to do some cloning and `npm link`'ing of the dependency chain, (specifically, performance openinsights and `insights.js` so chat with me if you wish to do this.

### TODO
- [x] Add CI
- [x] Flesh out a README.md
- [x] Update `@openinsights/openinsights` dependency once it has been published to NPM.
- [ ] Publish this as a module to NPM for consumption  in Insights.js

### CODE REVIEW

File: src/index.spec.ts [hunk 1]
```diff
@@ -0,0 +1,301 @@
...
+import entryFixture from "./fixtures/entry";
+
+const settingsFixture = configFixture.settings;
+const taskFixture = configFixture.tasks[1];
+
+describe("FastlyProvider", (): void => {
+  let providerFixture: Provider;
+
+  beforeEach((): void => {
+    providerFixture = new Provider(settingsFixture) as Provider;
```
Comment (psbanka) [inline src/index.spec.ts:19]: Odd. Why is it necessary to cast this as a `Provider`?

File: src/index.spec.ts [hunk 2]
```diff
@@ -0,0 +1,301 @@
...
+    providerFixture.setSessionConfig(configFixture);
+    fetchMock.resetMocks();
+  });
+
+  describe("shouldRun", (): void => {
+    it("should always run", () => {
+      const result = new Provider(settingsFixture);
```
Comment (psbanka) [inline src/index.spec.ts:26]: you already instantiate a `Provider` in the `beforeEach()` method. Why instantiate a new one here?

File: src/index.spec.ts [hunk 3]
```diff
@@ -0,0 +1,301 @@
...
+  describe("fetchSessionConfig", (): void => {
+    it("should fetch session config", (): Promise<void> => {
+      fetchMock.mockResponseOnce(JSON.stringify(configFixture), {
+        headers: {
+          "access-control-allow-origin": "*",
+        },
+      });
+      const provider = new Provider(settingsFixture);
+      return provider.fetchSessionConfig().then((result) => {
+        expect(result).toEqual(configFixture);
```
Comment (psbanka) [inline src/index.spec.ts:40]: is it not useful to test the URL endpoints for this request? e.g.
```
    expect(fetchMock.calls[0][0]).toEqual('https://api.fastly.com') // or whatever
```

File: src/index.spec.ts [hunk 4]
```diff
@@ -0,0 +1,301 @@
...
+      });
+    });
+  });
+
+  describe("expandTasks", (): void => {
+    mockRandomForEach([0.5]);
+    it("should expand session config to tasks", (): void => {
+      const result = providerFixture.expandTasks();
+      expect(result).toHaveLength(3);
```
Comment (psbanka) [inline src/index.spec.ts:63]: The magic number `3` is not very clear here. Can we just use `config.tasks.length` ?

File: src/index.ts [hunk 5]
```diff
@@ -0,0 +1,286 @@
+import {
+  Provider as ProviderBase,
```
Comment (psbanka) [inline src/index.ts:2]: Why the rename?

File: src/index.ts [hunk 6]
```diff
@@ -0,0 +1,286 @@
...
+  private _configUrl: string;
+  private _libraryVersion: string;
+  private _resourceCache: ResourceCache = {};
+
+  /**
+   * The main constructor; it recives the settings object the site owner
+   * initialised the provider with sets required properties such as the
+   * config URL.
+   *
+   * @param settings Settings to configure the provider
```
Comment (psbanka) [inline src/index.ts:43]: Do we need these parameter docs when we have typescript? I generally dispense with them, but that's cool if you're into it.

File: src/index.ts [hunk 7]
```diff
@@ -0,0 +1,286 @@
...
+   */
+  constructor(settings: Settings) {
+    // Call the ProviderBase constructor passing it the name of this provider.
```
Comment (psbanka) [inline src/index.ts:46]: is this comment necessary? seems pretty obvious to me what this is doing.

File: src/index.ts [hunk 8]
```diff
@@ -0,0 +1,286 @@
...
+    if (settings.library_version && settings.library_version !== "") {
+      this._libraryVersion = settings.library_version;
+    }
+  }
+
+  /**
+   * A hook called during initialisation, to determine whether the provider
+   * should participate in the session. We currently set this to be always true,
+   * however may eventually derive the answer from configuraiton settings.
+   */
```
Comment (psbanka) [inline src/index.ts:64]: all these doc-strings are excellent. nice work.

**Review by psbanka — APPROVED** [review]
Great work! Glad to see this and thanks for asking me to review

File: src/index.ts (hunk 6 above)
Comment (phamann) [reply src/index.ts:43]: A couple of reasons:
- Most of these param lists were copy/pasta from the openinsights provider class, which we use tsdoc to export proper documentation.
- This will allow us to export documentation for the provider if we want to in the future.
